### PR TITLE
Implementation of automotive OBD protocol

### PR DESCRIPTION
**Pull request by akorb**
This Pull-Request adds the automotive [On-board diagnostics protocol](https://en.wikipedia.org/wiki/OBD-II_PIDs). This protocol is used for mostly emission-relevant diagnostic of a car. Additionally, it can read even non-emission-relevant Diagnostic Trouble Codes.

It is usually used with the `ISOTPSocket`.

Since many of the PID classes are extremely similar, only differing in the binding value, not every single class is tested. But I added tests which test all **kind** of classes, meaning different lengths and different services.

### CODE REVIEW

**Comment by polybassa** [comment]
Thanks for this PR. Remark for the maintainers: Every class has a different kind of representation, for example the Engine RPM is calculated by the formula (256 * A + B) / 4 which is represented by this class in Scapy 
```
class Pid0C_S1AndS2(Packet):
    name = "PID_0C_EngineRpm"
    fields_desc = [
        StrFixedLenField('data', b'', 2)
    ]
```
At the moment, most of the data fields are "placeholders". We want to give the proper representation to this protocol later and piece by piece.

File: scapy/contrib/automotive/obd/pids_s1AndS2.py [hunk 1]
```diff
@@ -0,0 +1,1165 @@
...
+
+
+class Pid03_S1AndS2(Packet):
+    name = "PID_03_FuelSystemStatus"
+    fields_desc = [
+        StrFixedLenField('data', b'', 2)
+    ]
+
+
+class Pid04_S1AndS2(Packet):
```
Comment (gpotter2) [inline scapy/contrib/automotive/obd/pids_s1AndS2.py]: Hmm, I would rather have every class named by its name (below), which sounds clearer. If it is on purpose and has a special meaning, It’s fine

Comment (akorb) [reply scapy/contrib/automotive/obd/pids_s1AndS2.py]: The naming convention is explained in the doc:
https://github.com/secdev/scapy/pull/1614/commits/55c77e7be6350868a8b74f26b2d33c61aaec2987#diff-8026137693fcbeb88c642f33c81104cfR1590

Comment (gpotter2) [reply scapy/contrib/automotive/obd/pids_s1AndS2.py]: Ok got it

**Comment by polybassa** [comment]
Can you please add the obd folder path to the setup.py file.

**Comment by akorb** [comment]
@polybassa Done

**Comment by akorb** [comment]
@gpotter2 anything else to change?

**Comment by gpotter2** [comment]
Calling @guedou or @p-l-

I don’t really know what to think of the PR. It’s a cool feature but I generally don’t like having hundreds of different packets with IDs. Maybe it’s the only way

**Comment by akorb** [comment]
I understand. It's difficult, because each ID has a different payload.
Always open for suggestions.

**Comment by polybassa** [comment]
Every class needs also a custom representation, which has to be implemented in the future. So there are here mainly as placeholders at the moment.

**Comment by akorb** [comment]
I started implementiI started implementing pid specific fields. Thus the classes enlarge.

pids_s1AndS2.py has now ~1600 lines. This should be splitted. Files containing a specific range of pids would make sense.

My suggestion is:
image


Let me know if you think this is sufficient.

**Comment by polybassa** [comment]
I don't like the split to multiple files. This would require to load them all individually. I would prefer to have just one obd file, which can be loaded through `load_contrib('automotive.obd')`

**Comment by akorb** [comment]
The obd file would load all PIDs. But I understand your concerns.
You are fine with a file with ~1600 lines?

**Comment by polybassa** [comment]
I don’t see a problem in big files. Most of the stuff is just protocol definitions.

**Comment by gpotter2** [comment]
I must say that having a smaller set of files, as proposed by @akorb (as in the tree of the linked image), would be much cleaner than having a big file.

**Comment by akorb** [comment]
The obd.py file would import all PIDs.
Thus load_contrib('automotive.obd.obd') is still sufficient.

Shall I implement it this way? Then I would include this structural change with my next major commit.

**Comment by akorb** [comment]
Still in progress, Just pushed for a colleague and backup concerns.
No review needed. Some tests will fail.

File: scapy/contrib/automotive/obd/iid/iids.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    ]
+
+
+class OBD_IID04(Packet):
+    name = "IID_04_CalibrationId"
+    fields_desc = [
+        ByteField('messageCount', 1),
+        FieldListField('calibrationIdentifications', None,
+                       StrFixedLenField(b'', 0, 16),
+                       count_from=lambda pkt: len(pkt.original) / 16)
```
Comment (gpotter2) [inline scapy/contrib/automotive/obd/iid/iids.py]: It would be slightly cleaner to use a `//` with `count_from`. The API isn’t really thought for decimal values

**Comment by gpotter2** [comment]
You have lots of PEP8 reports:
https://travis-ci.org/secdev/scapy/jobs/467962773

It’s obviously cleaner to use a star import here, knowing the amount of things to import, so I would say adding a global ignore on top of the file

```
# flake8:  Fxxx
```
Is justified

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py [hunk 3]
```diff
@@ -7,6 +7,12 @@
 from scapy.fields import StrFixedLenField, BitEnumField, BitField, \
     ScalingField, ConditionalField, FlagsField, XByteEnumField, ShortField
 from scapy.packet import Packet
+import scapy.modules.six as six
+
+if six.PY2:
+    _temperature = "\xC2\xB0C"
+else:
+    _temperature = "\xB0C"
```
[the remark this answers is not in the document]
Comment (akorb) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]:
```
>>> '\xB0C'
'°C'
```

It works with python3.

But this does not:
```
>>> b'\xB0C'
b'\xb0C'
```

What is your suggestion?

[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: You are right, your change is actually quite smart

File: scapy/fields.py [hunk 4]
```diff
@@ -2128,3 +2128,52 @@ def i2repr(self, pkt, x):
...
+        return x
+
+    def i2repr(self, pkt, x):
+        return "%s %s" % (self.i2h(pkt, x), self.unit)
+
+
+class ScalingIntField(ScalingField):
+    # Avoid roundings to prevent precision loss
+    def __init__(self, name, default, scaling=1, unit="",
+                 offset=0, fmt="B"):
```
Comment (polybassa) [inline scapy/fields.py]: The name `ScalingIntField` is easy to misunderstand. All `IntFields` in Scapy dissect usually 4 bytes. This isn't the case for this Field. Can you give me an example, where you lose the precision through rounding? If you want to prevent `round` from producing fractional digits, you can use the parameter `ndigits=0`.

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (polybassa) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: Thanks. Do you see a better place for this code? Since I repeat it on all the PID files.

File: scapy/fields.py (hunk 4 above)
Comment (akorb) [reply scapy/fields.py]: I agree with you. The name is misleading.

The problem is not the round, but the conversion to `float` in general.
Example:
```
>>> 18446744073709551615 == 18446744073709551615.0
False
```
This happens for very high numbers (long). OBD_PID7F has three of those long fields (8 bytes).
This is also a potential solution:
`ScalingField('total', 0, unit='sec', fmt='Q', offset=0, ndigits=0, scaling=1),`
Since it sets the default float values to integer values and thus prevents converting to float. But it feels like a hacky workaround.

Suggestions?

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (akorb) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: The python2 version does not work for Windows.
```python
_temperature = u'\N{DEGREE SIGN}C'
```
Works for me with python 2&3 on Windows and Linux

File: scapy/fields.py (hunk 4 above)
Comment (polybassa) [reply scapy/fields.py]: I pushed an commit

Comment (akorb) [reply scapy/fields.py]: Does not work, because:
```
>>> 5 - 3.0
2.0
```

And `offset` is a float by default. `offset` has to be `0` (no float!) too.
Maybe it's the best to set the default values from `offset` to `0` and for `scaling` to `1`, instead of requiring to pass the integer value to the constructor.

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (polybassa) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: Well, that doesnt work on python2 Linux for me

File: scapy/fields.py (hunk 4 above)
Comment (polybassa) [reply scapy/fields.py]: Ok, I'll think about the current implementation, and may refactor it in the next days

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (akorb) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]:
```bash
~ $ python2
Python 2.7.15 (default, Jun 27 2018, 13:05:28) 
[GCC 8.1.1 20180531] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> _temperature = u'\N{DEGREE SIGN}C'
>>> print(_temperature)
°C
```
Confusing. Works fine for me

[the remark this answers is not in the document]
Comment (polybassa) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: I double checked it. You have to test the `show` function in scapy. On my system, this function produces a different output than simple `print(u'\N{DEGREE SIGN}C')` tests. I add another case for windows in the current implementation.

File: scapy/fields.py (hunk 4 above)
Comment (polybassa) [reply scapy/fields.py]: Another example to think about. What behaviour would you expect? A or B?
```
>>> class DebugPacket(Packet):
...:    fields_desc = [
...:        ScalingField('data', 0, scaling=10)
...:    ]
...:
...:x = DebugPacket()
...:
...:x.data = 0x01
...:print(x.data)
A: 10.0
B: 1
>>> bytes(x)
A: '\x01'
B: '\x00'       => because of rounding
>>> x.data = 1.0
>>> print(x.data)
A: 1.0
B: 1
>>> bytes(x)
A: '\x00'       => because of rounding
B: '\x00'       => because of rounding
>>> x.data = 10
>>> print(x.data)
A: 100.0
B: 10
>>> bytes(x)
A: '\x0A'
B: '\x01' 
>>> x.data = 10.0
>>> print(x.data)
A: 10.0
B: 10
>>> bytes(x)
A: '\x01'
B: '\x01'      
>>> x.data = 0x0f
>>> print(x.data)
A: 150.0
B: 15
>>> bytes(x)
A: '\x0f'
B: '\x02'    => because of rounding
```

Comment (polybassa) [reply scapy/fields.py]: At the moment, the `ScalingField` implements behaviour A. If we want to have Integers as internal storage, we should consider an implementation which behaves like B.

Comment (akorb) [reply scapy/fields.py]: That's a very good question.
I feel like the best option is to work always with the "raw" values, and only use the scaling in the `show` methods. Otherwise this would be very confusing. I actually needed a few minutes before I understood the output of A in the first example. Thus I would expect B in the most examples you provided.

Only for the most, because:

```python
>>> x.data = 10
>>> print(x.data)
A: 100.0
B: 10
>>> bytes(x)
A: '\x0A'
B: '\x01' 
```

In the first one I would expect B. But in the second one I would expect A, no matter if `data` is a float or an integer.


In summary, my idea is: The field contains the **data**, the `show` methods display the **information**. `bytes(x)` shall contain the same value as in the properties (like `data`).
`ScalingField` might provide a method which provides the implementation of A, thus converting the actual information to the necessary data.

Does this make sense to you and is it even implementable?

**Comment by polybassa** [comment]
> Could you make that unique ? If you rename _temperature to something less confusing (e.g `_celcius_unit` for instance), you can even add it to `scapy.compat`

Of course. Thanks that you pointed me to `scapy.compat`

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (polybassa) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: @akorb Could you please double check, that the current implementation works on windows.

File: scapy/fields.py (hunk 4 above)
Comment (polybassa) [reply scapy/fields.py]: Your summary makes sense. There are advantages and disadvantages of both strategies (using float vs. raw values internally). In my opinion, the **information value** is more important than the **raw value**. I think, that is also the present philosophy of Scapy. A big disadvantage when using raw values internally is the assignment of information values. We would need to create a complete new method for the conversion of an information value to a raw value, in order to be able to assign that value. 

Here is a list of features, we have to consider for ScalingFields:

- Assignment of information values `x.data = 150.0'
- Assignment of raw values `x.data = 0x0f` or `x.data = 15`
- Representation of information values with proper units
- Handling of signed information values
- Handling of information values with offset

File: scapy/contrib/automotive/obd/pid/pids_00_1F.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (akorb) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: The current implementation works. I also checked your former implementation with `show` instead of `print` and this works too. Thus this actually already worked for windows with `show`.
```python
if six.PY2:
    _temperature = "\xC2\xB0C"
else:
    _temperature = "\xB0C"
```

Sorry about that.

[the remark this answers is not in the document]
Comment (akorb) [reply scapy/contrib/automotive/obd/pid/pids_00_1F.py]: For python3 it works. Python2 throws an error:

```
>>> OBD()/OBD_S01()/OBD_PID05(data=83)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "scapy\packet.py", line 435, in __repr__
    repr(self.payload),
  File "scapy\packet.py", line 435, in __repr__
    repr(self.payload),
  File "scapy\packet.py", line 430, in __repr__
    vcol(val))
  File "scapy\themes.py", line 53, in do_style
    val = str(val)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xb0' in position 5: ordinal not in range(128)
```

This might help: https://stackoverflow.com/a/9942822/2050020

**Comment by polybassa** [comment]
> I'm not sure we really want to do that in Scapy. Cannot you write "Celsius" or something else?
> 
> BTW I'm not sure about the LINUX test: how does it work with *BSD, MacOS, etc.?
> 
> Also FYI, under Linux, `u"\xb0c"` works with both Python 2 & 3 here.
I've checked `u"\xb0c"` with this code:
```
>>> load_contrib("automotive.obd.obd")
...:resp2 = OBD(service=65)/OBD_S01(pid=5)/OBD_PID05(data=41.0)
...:resp2.show()
```
Python2 gave me this error: 
```
UnicodeEncodeError: 'ascii' codec can't encode character u'\xb0' in position 5: ordinal not in range(128)
```

File: scapy/fields.py (hunk 4 above)
Comment (polybassa) [reply scapy/fields.py]: I'm pretty happy with the latest implementation. 
Last open question. Should we use a different name for this field. I'm not so happy with ScalingField. 
@Maintainers: How should we name this field? Maybe ScaledField or UnitField or ValueField, or something completely different.

**Comment by p-l-** [comment]
@polybassa I don't think Scapy should output non-ASCII chars by default, if possible. That will bring a lot of bugs, since it will depend (under Unix systems) on the terminal settings and the locale environment variables that we do not want to handle.

File: scapy/contrib/automotive/obd/obd.py [hunk 5]
```diff
@@ -62,13 +62,14 @@ def answers(self, other):
...
 # Service Bindings
+
 bind_layers(OBD, OBD_S01, service=0x01)
 bind_layers(OBD, OBD_S02, service=0x02)
 bind_layers(OBD, OBD_S03, service=0x03)
 bind_layers(OBD, OBD_S04, service=0x04)
 bind_layers(OBD, OBD_S06, service=0x06)
 bind_layers(OBD, OBD_S07, service=0x07)
-bind_layers(OBD, OBD_S08, service=0x08)
+bind_layers(OBD, OBD_TID_Record, service=0x08)
```
Comment (akorb) [inline scapy/contrib/automotive/obd/obd.py]: Isn't it weird that each class name contains its service number but `OBD_TID_Record `does not?
Is there are reason for this?
The same for `OBD_MID `and `OBD_TID `

Comment (polybassa) [reply scapy/contrib/automotive/obd/obd.py]: I agree. I changed it.

**Comment by gpotter2** [comment]
Hello @polybossa @akorb,
We’re going to do an “emergency” release in the following days, as we were notified of a (minor) security issue. 

What’s the state of the OBD PRs ? Are you fine being missed by 2.4.1 or do you think this is ready to be merged ?

**Comment by akorb** [comment]
> 
> 
> Hello @polybassa @akorb,
> We’re going to do an “emergency” release in the following days, as we were notified of a (minor) security issue.
> 
> What’s the state of the OBD PRs ? Are you fine being missed by 2.4.1 or do you think this is ready to be merged ?

I prefer not beeing merged.

**Comment by polybassa** [comment]
I would say, the PR almost finish, but we want to do some final checks and add more unit tests in the next days. Its fine not being merged.

**Comment by gpotter2** [comment]
It’s fine, we’ll probably get 2.4.2 in a matter of weeks. That leaves you all the time you may need :)

**Comment by gpotter2** [comment]
Please rather us `git pull --rebase upstream master && git push -f`, to avoid the merge commit.

This PR will most likely be squashed

**Comment by polybassa** [comment]
Sorry for this inconvenience here. This PR got out of hand.

**Comment by polybassa** [comment]
@akorb: Can you please close this PR. Since the development of the OBD layer is almost done, except some documentation, we should switch to the squased branche at PR #1787

**Comment by guedou** [comment]
@akorb @polybassa Done =)
